docs/quick-release-notes/main.go

```diff
@@ -0,0 +1,167 @@
+	Commit     struct {
+		SHA string `json:"sha"`
+		URL string `json:"url"`
+	} `json:"commit"`
+	NodeID string `json:"node_id"`
+}
+
+func getTagsFromAPI(owner string, repo string) []Tag {
+	// https://docs.github.com/en/rest/reference/repos#list-repository-tags
+	pathToTags := fmt.Sprintf("https://api.github.com/repos/%v/%v/tags", owner, repo)
+	resp, err := http.Get(pathToTags)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	sb := string(body)
+	var tagsForEachRelease []Tag
+	json.Unmarshal([]byte(sb), &tagsForEachRelease)
+	return tagsForEachRelease
+}
+
+func getChangeLogInfo(scanner *bufio.Scanner) ([]int, []string, int, []string) {
+	// regexp for: ## 1.0.0 (2021-12-13)
+	r, err := regexp.Compile(`## \d+\.\d+\.\d+`)
+	if (err) != nil {
+		log.Fatal(err)
+	}
+	lineCount := 0
+	counterForChangeLogLength := 0
+	var versionLineNumbers []int
+	var changeLogText []string
+	var versionTitlesForEachRelease []string
+	for scanner.Scan() {
+		counterForChangeLogLength++
+		changeLogText = append(changeLogText, scanner.Text())
+		if r.MatchString(scanner.Text()) {
+			// ## 2.3.0 (2022-02-24) => 2.3.0 (2022-02-24)
+			versionNumberAndDateWithoutHashes := scanner.Text()[3:]
+			versionTitlesForEachRelease = append(versionTitlesForEachRelease, versionNumberAndDateWithoutHashes)
+			versionLineNumbers = append(versionLineNumbers, lineCount)
+		}
+		lineCount++
+	}
+	if scanner.Err() != nil {
+		fmt.Println(scanner.Err())
+	}
+	return versionLineNumbers, changeLogText, counterForChangeLogLength, versionTitlesForEachRelease
+}
+
+func makeReleaseSlice(versionLineNumbers []int, changeLogText []string, changeLogLength int) []string {
+	// Store contributors and errors from end of changelog.
+	contributorsAndErrors := ""
+	numberOfLinesForLastRelease := 5
+	for i := numberOfLinesForLastRelease + versionLineNumbers[len(versionLineNumbers)-1]; i < changeLogLength; i++ {
+		contributorsAndErrors += changeLogText[i] + "\n"
+	}
+	var releaseNotesForEachVersion []string
+	// exclude Last version (## 0.2.0) with - 1
+	versionLineNumbersLengthMinusOne := len(versionLineNumbers) - 1
+	for i, versionLineNumber := range versionLineNumbers[:] {
+		releaseBodyLines := ""
+		if i < (versionLineNumbersLengthMinusOne) {
+			for j := versionLineNumbers[i] + 1; j < versionLineNumbers[i+1]; j++ {
+				releaseBodyLines += changeLogText[j] + "\n"
+			}
+		}
+		// Handle last version (## 0.2.0)
+		if versionLineNumber == versionLineNumbers[versionLineNumbersLengthMinusOne] {
+			for j := 1; j < numberOfLinesForLastRelease; j++ {
+				releaseBodyLines += changeLogText[versionLineNumber+j] + "\n"
+			}
+		}
+		releaseNotesForEachVersion = append(releaseNotesForEachVersion, releaseBodyLines+contributorsAndErrors)
+	}
+	return releaseNotesForEachVersion
+}
+
+func sendToGitHubAPI(tagForRelease Tag, releaseNote string, versionTitle string, owner string, repo string) {
```

Comment:
Nit: Rename this function to describe what it accomplishes. Perhaps `addOrUpdateGitHubRelease`?

Reply:
It's not a `http` `Patch` request though it doesn't `Update` them I called it: `makeGitHubRelease`

Reply:
> it doesn't `Update` them

I thought you said it is idempotent: https://github.com/quick-lint/quick-lint-js/pull/669#issuecomment-1068888765

Reply:
Maybe I have a different understanding of idempotent. Putting the fancy terminology aside. It does not duplicate the releases if they already exist, nor does it overwrite/update them. Therefore it only makes them if they don't already exist. Just like if a user clicks a `pay now` button on an e-commerce site multiple times it only pays for it once; that is my understanding of idempotent..

Reply:
My bad. I think I meant something else when I said 'idempotent'.

I want the following behavior:
* If I run the script twice, the second run should have no effect. (The second run should notice that release were already made. It shouldn't create new GitHub releases.) (It sounds like this is implemented.)
* If I create a tag and an entry in the changelog for that tag, then I run the script, the script should create one new GitHub release. (It sounds like this is implemented.)
* If I edit the changelog for an existing tag/version/release, then I run the script, the script should modify the existing GitHub release. (It sounds like this is **not** implemented.)